source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,733 @@
+/++
+This module helps building functions from other functions.
+
+ It is sometimes necessary to create a function which is an exact copy of
+ another function. Or sometimes it is necessary to introduce a few variations,
+ while carrying all the other aspects. Because of function attributes,
+ parameter storage classes and user-defined attributes, this requires building
+ a string mixin. In addition, the mixed-in code must refer only to local names,
+ if it is to work across module boundaires. This problem and its solution are
+ by Adam D. Ruppe in a Tip of the Week, available here:
+ https://stackoverflow.com/questions/32615733/struct-composition-with-mixin-and-templates/32621854#32621854
+
+ This module facilitates the  creation of such mixins.
+
+ +/
+
+module bolts.experimental.refraction;
+
+import std.array;
+import std.format;
+import std.meta;
+import std.algorithm.iteration : map;
+import std.range : iota;
+import std.traits : functionAttributes, FunctionAttribute;
+
+// Do not require caller module to import 'std.traits'. Instead use our own
+// aliases in mixtures.
+alias ReturnType = std.traits.ReturnType;
+alias Parameters = std.traits.Parameters;
+
+template ParameterAttribute(alias F, int i, int j) {
+    static if (is(typeof(F) P == __parameters)) {
+        alias ParameterAttribute = Alias!(__traits(getAttributes, P[i..i+1])[j]);
+    }
+}
+
+unittest {
+    struct virtual;
+    void kick(int times, @virtual @("Animal") Object animal);
+    static assert(is(ParameterAttribute!(kick, 1, 0) == virtual));
+    static assert(ParameterAttribute!(kick, 1, 1) == "Animal");
+}
+
+/**
+   Return a `Function` object that captures all the aspects of `fun`, using the
+   value of `localName` to represent the return and parameter types, and the
+   UDAs.
+
+   The `localName` parameter is, in general, *not* the function name. Rather,
+   it is a compile-time expression that involves only symbols that exist in the
+   caller's scope, for example a function alias passed as a template
+   parameter. See
+   https://stackoverflow.com/questions/32615733/struct-composition-with-mixin-and-templates/32621854#32621854
+   for a detailed explanation.
+
+   Params:
+   fun = a function
+   localName = a string that represents `fun` in the caller's context
+*/
+
+Function refract(alias fun, string localName)()
```

Comment:
> Why not have this one take an overload index as well?

Indeed why not, and give it a default value of -1:

```d
Function refract(alias fun, string localName, uint index = -1)()
```

 It would make it possible to write `refract!(foo, "foo", 2)` directly, instead of `refract!(foo, "foo").withIndex(2)`. In fact if we do this, we can remove `withIndex` altogether, as your remark above ("What's the use of changing the index then?") suggests. Either the user needs to store a flat, module-global index, or an overload set local index, but once it's set, there's no changing it, because functions don't wander in modules.

Yeah, thinking about it, it looks better like this...